src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -266,72 +266,99 @@ protected BoundExpression LowerTest(BoundDagTest test)
 
                     case BoundDagValueTest d:
                         Debug.Assert(!input.Type.IsNullableType());
-                        return MakeEqual(_localRewriter.MakeLiteral(d.Syntax, d.Value, input.Type), input);
+                        return makeEqualityTest(_localRewriter.MakeLiteral(d.Syntax, d.Value, input.Type), input);
+
+                    case BoundDagRelationalTest d:
+                        Debug.Assert(!input.Type.IsNullableType());
+                        Debug.Assert(input.Type.IsValueType);
+                        return makeRelationalTest(input, d.OperatorKind, _localRewriter.MakeLiteral(d.Syntax, d.Value, input.Type));
 
                     default:
                         throw ExceptionUtilities.UnexpectedValue(test);
                 }
-            }
 
-            private BoundExpression MakeEqual(BoundExpression loweredLiteral, BoundExpression input)
-            {
-                Debug.Assert(loweredLiteral.Type.Equals(input.Type, TypeCompareKind.AllIgnoreOptions));
-
-                if (loweredLiteral.Type.SpecialType == SpecialType.System_Double && double.IsNaN(loweredLiteral.ConstantValue.DoubleValue))
-                {
-                    // produce double.IsNaN(input)
-                    return _factory.StaticCall(SpecialMember.System_Double__IsNaN, input);
-                }
-                else if (loweredLiteral.Type.SpecialType == SpecialType.System_Single && float.IsNaN(loweredLiteral.ConstantValue.SingleValue))
+                BoundExpression makeRelationalTest(BoundExpression input, BinaryOperatorKind operatorKind, BoundExpression literal)
                 {
-                    // produce float.IsNaN(input)
-                    return _factory.StaticCall(SpecialMember.System_Single__IsNaN, input);
+                    TypeSymbol comparisonType = input.Type.EnumUnderlyingTypeOrSelf();
+                    if (operatorKind.OperandTypes() == BinaryOperatorKind.Int && comparisonType.SpecialType != SpecialType.System_Int32)
+                    {
+                        // Promote operands to int before comparison for byte, sbyte, short, ushort
+                        Debug.Assert(comparisonType.SpecialType switch
+                        {
+                            SpecialType.System_Byte => true,
+                            SpecialType.System_SByte => true,
+                            SpecialType.System_Int16 => true,
+                            SpecialType.System_UInt16 => true,
+                            _ => false
```

Comment:
>_ [](start = 28, length = 1)

Do we get here for `char`?

Reply:
Values of type `char` are not compared by promotion to `int`, so no, we don't get here.

---
In reply to: [383043865](https://github.com/dotnet/roslyn/pull/41503#discussion_r383043865) [](ancestors = 383043865)